Plutarch/Lift.hs

```diff
@@ -1,109 +1,70 @@
+{-# LANGUAGE AllowAmbiguousTypes #-}
+{-# LANGUAGE FunctionalDependencies #-}
 {-# LANGUAGE UndecidableInstances #-}
+{-# LANGUAGE UndecidableSuperClasses #-}
 
-module Plutarch.Lift (
-  -- * Converstion between Plutarch terms and Haskell types
-  pconstant,
-  plift,
-  LiftError (..),
+module Plutarch.Lift (PConstant (..), PUnsafeLiftDecl (..), PLift, pconstant, plift, plift', LiftError, DerivePConstantViaCoercible (..), DerivePConstantViaNewtype (..)) where
 
-  -- * Define your own conversion
-  PLift (..),
-
-  -- * Internal use
-  PBuiltinType (..),
-) where
-
-import Data.Bifunctor (first)
-import Data.Data (Proxy (Proxy))
+import Data.Coerce
 import Data.Kind (Type)
-import Data.String
-import Data.Text
-import qualified Data.Text as T
 import GHC.Stack (HasCallStack)
 import Plutarch.Evaluate (evaluateScript)
-import Plutarch.Internal (
-  ClosedTerm,
-  Term,
-  compile,
-  punsafeConstantInternal,
- )
+import Plutarch.Internal (ClosedTerm, PType, Term, compile, punsafeConstantInternal)
 import qualified Plutus.V1.Ledger.Scripts as Scripts
 import qualified PlutusCore as PLC
-import PlutusCore.Constant (readKnownSelf)
-import qualified PlutusCore.Constant as PLC
-import PlutusCore.Evaluation.Machine.Exception (
-  EvaluationException,
-  MachineError,
- )
+import PlutusCore.Constant (readKnownConstant)
+import PlutusCore.Evaluation.Machine.Exception (ErrorWithCause, MachineError)
 import qualified UntypedPlutusCore as UPLC
-import UntypedPlutusCore.Evaluation.Machine.Cek (CekUserError)
-
--- | Error during script evaluation.
-data LiftError
-  = LiftError_ScriptError Scripts.ScriptError
-  | LiftError_EvalException T.Text -- Using Text, because there is no Eq possible with DeBruijn naming.
-  | LiftError_Custom T.Text
-  deriving stock (Eq, Show)
-
-instance IsString LiftError where
-  fromString = LiftError_Custom . T.pack
-
-{- | Class of Plutarch types `p` that can be converted to/from a Haskell type.
 
-The Haskell type is determined by `PHaskellType p`.
--}
-class PLift (p :: k -> Type) where
-  -- | The associated Haskell type for `p`
-  type PHaskellType p :: Type
+class (PConstant (PLifted p), PConstanted (PLifted p) ~ p) => PUnsafeLiftDecl (p :: PType) where
```

Comment:
Would an open type family make more sense here? It's just an empty class with an associated type after all.

Reply:
Ah, right... darn.